src/v2/view-builder/views/DeviceEnrollmentTerminalView.js

```diff
@@ -0,0 +1,60 @@
+import { loc } from 'okta';
+import BaseHeader from '../internals/BaseHeader';
+import HeaderBeacon from '../components/HeaderBeacon';
+import BaseView from '../internals/BaseView';
+import BaseForm from '../internals/BaseForm';
+import OdaOktaVerifyTerminalView from '../components/OdaOktaVerifyTerminalView';
+import MdmOktaVerifyTerminalView from '../components/MdmOktaVerifyTerminalView';
+
+const ODA = 'oda';
+const MDM = 'mdm';
+
+const Header = BaseHeader.extend({
+  HeaderBeacon: HeaderBeacon.extend({
+    getBeaconClassName: () => 'mfa-okta-verify',
+  }),
+  initialize () {
+    const deviceEnrollment = this.options.appState.get('deviceEnrollment');
+    this.enrollmentType = (deviceEnrollment.name || '').toLowerCase(); // oda/mdm
+    if (this.enrollmentType === ODA) { // add HeaderBeacon only for ODA
+      BaseHeader.prototype.initialize.apply(this, arguments);
+    }
+  },
+  postRender () {
+    if (this.enrollmentType === ODA) { // show HeaderBeacon only for ODA
+      BaseHeader.prototype.postRender.apply(this, arguments);
+    }
+  },
+});
+
+const Body = BaseForm.extend({
+  noButtonBar: true,
+
+  className: 'device-enrollment-terminal',
+
+  title () {
+    return this.enrollmentType === ODA
+      ? loc('enroll.title.oda', 'login')
+      : loc('enroll.title.mdm', 'login');
+  },
+
+  initialize () {
+    BaseForm.prototype.initialize.apply(this, arguments);
+    const deviceEnrollment = this.options.appState.get('deviceEnrollment');
+    this.enrollmentType = (deviceEnrollment.name || '').toLowerCase(); // oda/mdm
+    switch (this.enrollmentType) {
+    case ODA:
+      this.add(OdaOktaVerifyTerminalView);
+      break;
+    case MDM:
+      this.add(MdmOktaVerifyTerminalView);
+      break;
+    }
+  },
+});
+
+export default BaseView.extend({
+  Header,
```

Comment:
You can change the BaseView to support `Header` can be value or function, then you could do
```
Header() {
  if is ODA return ODAHeader
  else return null;
}

```

Reply:
What type of this value should be? Any View is a function.
And returning null is an unexpected behavior. I assume `Header` property should return View in any case.

We have to change HeaderBeacon for MDM anyway. That would add more checks here.

What is your purpose?

Reply:
@ivandenysenko-okta i agree with @haishengwu-okta, Your `const Header` implementation is basically duplicating what `BaseHeader` does.. Here we have opportunity to improve `Header` in BaseView to support such cases and generalize it.

Reply:
So, how do we check if Header is a function or a View?

Usage of _.result(this, 'Header'); fires this.Header() right?
But as long as it is a constructor, we can’t do that.
Particularly when we render a view where Header is a View it will throw an Error
<img width="1110" alt="Screen Shot 2020-10-28 at 6 42 04 PM" src="https://user-images.githubusercontent.com/54867795/97476901-a372f000-1957-11eb-8d0d-b339773d0ee6.png">

Maybe I totally misunderstand what is your idea.
Can you please give a working example to be more clear?

Reply:
I got this and looks like work for me. can you push your changes if it still doesn't work.

```
2 files changed, 25 insertions(+), 23 deletions(-)
src/v2/view-builder/internals/BaseView.js          | 12 +++++---
.../views/DeviceEnrollmentTerminalView.js          | 36 ++++++++++------------

modified   src/v2/view-builder/internals/BaseView.js
@@ -42,10 +42,14 @@ export default View.extend({
   preRender () {
     View.prototype.preRender.apply(this, arguments);
     // Add Views
-    this.add(this.Header, {
-      selector: '.siw-main-header',
-      options: this.options,
-    });
+    //
+    const HeaderView = _.result(this, 'Header');
+    if (HeaderView) {
+      this.add(HeaderView, {
+        selector: '.siw-main-header',
+        options: this.options,
+      });
+    }
     this.renderForm();
     this.add(this.Footer, {
       selector : '.siw-main-footer',
modified   src/v2/view-builder/views/DeviceEnrollmentTerminalView.js
@@ -11,18 +11,6 @@ const Header = BaseHeader.extend({
   HeaderBeacon: HeaderBeacon.extend({
     getBeaconClassName: () => 'mfa-okta-verify',
   }),
-  initialize () {
-    const deviceEnrollment = this.options.appState.get('deviceEnrollment');
-    this.enrollmentType = (deviceEnrollment.name || '').toLowerCase(); // oda/mdm
-    if (this.enrollmentType === Enums.ODA) { // add HeaderBeacon only for ODA
-      BaseHeader.prototype.initialize.apply(this, arguments);
-    }
-  },
-  postRender () {
-    if (this.enrollmentType === Enums.ODA) { // show HeaderBeacon only for ODA
-      BaseHeader.prototype.postRender.apply(this, arguments);
-    }
-  },
 });
 
 const Body = BaseForm.extend({
@@ -41,17 +29,27 @@ const Body = BaseForm.extend({
     const deviceEnrollment = this.options.appState.get('deviceEnrollment');
     this.enrollmentType = (deviceEnrollment.name || '').toLowerCase(); // oda/mdm
     switch (this.enrollmentType) {
-    case Enums.ODA:
-      this.add(OdaOktaVerifyTerminalView);
-      break;
-    case Enums.MDM:
-      this.add(MdmOktaVerifyTerminalView);
-      break;
+      case Enums.ODA:
+        this.add(OdaOktaVerifyTerminalView);
+        break;
+      case Enums.MDM:
+        this.add(MdmOktaVerifyTerminalView);
+        break;
     }
   },
 });
 
 export default BaseView.extend({
-  Header,
+  Header() {
+    const deviceEnrollment = this.options.appState.get('deviceEnrollment');
+    this.enrollmentType = (deviceEnrollment.name || '').toLowerCase(); // oda/mdm
+    //
+    if (this.enrollmentType === Enums.ODA) { // add HeaderBeacon only for ODA
+      return Header;
+    } else {
+      return null;
+    }
+
+  },
   Body,
 });
```

Reply:
Did you check it with other UIs?
I have a demo of how it works for me https://okta.box.com/s/0hmi82tpwcvclyrjuyblwz1fem5fd2au

Reply:
That was incorrect.
We can try to override `BaseView.initialize` to create `Header` properly. (details in email). LMK how it goes

Reply:
Added `initialize` with redefining `this.Header`